io-engine/src/grpc/v1/replica.rs

```diff
@@ -492,4 +573,99 @@ impl ReplicaRpc for ReplicaService {
         .await
     }
+    #[named]
+    async fn list_replica_snapshot(
+        &self,
+        request: Request<ListReplicaSnapshotsRequest>,
+    ) -> GrpcResult<ListReplicaSnapshotsResponse> {
+        self.locked(
+            GrpcClientContext::new(&request, function_name!()),
+            async move {
+                let args = request.into_inner();
+                info!("{:?}", args);
+                let rx = rpc_submit(async move {
+                    // if replica_uuid is valid, filter snapshot based on
+                    // replica_uuid
+                    if !args.replica_uuid.is_empty() {
+                        let lvol = match UntypedBdev::lookup_by_uuid_str(
+                            &args.replica_uuid,
+                        ) {
+                            Some(bdev) => Lvol::try_from(bdev)?,
+                            None => {
+                                return Err(LvsError::Invalid {
+                                    source: Errno::ENOENT,
+                                    msg: format!(
+                                        "Replica {} not found",
+                                        args.replica_uuid
+                                    ),
+                                })
+                            }
+                        };
+                        let snapshots = lvol
+                            .list_snapshot()
+                            .into_iter()
+                            .map(ReplicaSnapshot::from)
+                            .collect();
+                        Ok(ListReplicaSnapshotsResponse {
+                            snapshots,
+                        })
+                    } else {
+                        let snapshots = Lvol::list_all_snapshots()
+                            .into_iter()
+                            .map(ReplicaSnapshot::from)
+                            .collect();
+                        Ok(ListReplicaSnapshotsResponse {
+                            snapshots,
+                        })
+                    }
+                })?;
+                rx.await
+                    .map_err(|_| Status::cancelled("cancelled"))?
+                    .map_err(Status::from)
+                    .map(Response::new)
+            },
+        )
+        .await
+    }
+
+    #[named]
+    async fn delete_replica_snapshot(
+        &self,
+        request: Request<DeleteReplicaSnapshotRequest>,
+    ) -> GrpcResult<()> {
+        self.locked(
+            GrpcClientContext::new(&request, function_name!()),
+            async move {
+                let args = request.into_inner();
+                info!("{:?}", args);
+                let rx = rpc_submit(async move {
+                    let bdev = UntypedBdev::bdev_first().expect("Failed to enumerate devices");
+
+                    let mut devices = bdev
+                        .into_iter()
+                        .filter(|b| b.driver() == "lvol" && b.uuid_as_string() == args.snapshot_uuid)
```

Comment:
Rather than filter you could use .find which returns an Option rather than a list, then you don't have to handle multi instances found, which shouldn't really happen.

Reply:
Thank you for the suggestion.  Never observed difference between filter and find :) 
Changed the code.